bee-network/src/handshake.rs

```diff
@@ -0,0 +1,395 @@
+
+            let packet_type = packet.ty().map_err(HandshakeError::PacketType)?;
+
+            if matches!(packet_type, PacketType::Handshake) {
+                log::info!("received handshake request.");
+
+                let req = HandshakeRequest::from_protobuf(packet.data())?;
+                // println!("{:#?}", req);
+
+                let peer_req_data = req.protobuf()?;
+
+                HandshakeValidator::validate_request(&peer_req_data)?;
+                // println!("Received valid handshake request.");
+
+                send_handshake_response(writer, &peer_req_data, local_id).await?;
+
+                break packet;
+            }
+        }
+    };
+
+    let peer_public_key = packet.public_key();
+    let mut pk = [0u8; 32];
+    pk.copy_from_slice(&peer_public_key[..32]);
+    let peer_public_key = ed25519::PublicKey::try_from_bytes(pk).map_err(HandshakeError::PublicKey)?;
+    let peer_identity = Identity::from_public_key(peer_public_key);
+
+    Ok(peer_identity)
+}
+
+async fn await_response(
+    reader: &mut BufReader<OwnedReadHalf>,
+    local_req_data: BytesMut,
+    _peer_info: &ManualPeerInfo,
+) -> Result<Identity, HandshakeError> {
+    let mut buf = vec![0; MAX_HANDSHAKE_PACKET_SIZE];
+
+    let packet = loop {
+        if let Ok(num_received) = reader.read(&mut buf).await {
+            if num_received == 0 {
+                return Err(HandshakeError::ConnectionResetByPeer);
+            }
+
+            if num_received > MAX_HANDSHAKE_PACKET_SIZE {
+                return Err(HandshakeError::PacketSizeMismatch {
+                    received: num_received,
+                    max_allowed: MAX_HANDSHAKE_PACKET_SIZE,
+                });
+            }
+            // println!("Received {} bytes: {:?}", num_received, &buf[..num_received]);
+
+            let packet = Packet::from_protobuf(&buf[..num_received]).map_err(HandshakeError::Decode)?;
+            // println!("{:#?}", packet);
+
+            let packet_type = packet.ty().map_err(HandshakeError::PacketType)?;
+
+            if matches!(packet_type, PacketType::Handshake) {
+                log::info!("received handshake response.");
+
+                let res = HandshakeResponse::from_protobuf(packet.data())?;
+                // println!("{:#?}", res);
+
+                let peer_res_data = res.protobuf()?;
+
+                HandshakeValidator::validate_response(&peer_res_data, &local_req_data)?;
+
+                log::debug!("handshake response is valid");
+
+                break packet;
+            }
+        }
+    };
+
+    let peer_public_key = packet.public_key();
+    let mut pk = [0u8; 32];
+    pk.copy_from_slice(&peer_public_key[..32]);
+    let peer_public_key = ed25519::PublicKey::try_from_bytes(pk).map_err(HandshakeError::PublicKey)?;
+    let peer_identity = Identity::from_public_key(peer_public_key);
+
+    Ok(peer_identity)
+}
+
+/// Errors, that may occur during the handshake process.
+#[derive(Debug)]
+pub enum HandshakeError {
```

Comment:
What happened to `thiserror`? :trollface: (I'm just being annoying, I'm not bothered)

Reply:
on it ... finally :trollface: